raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -3426,6 +3428,12 @@ public Optional<SnapshotWriter<T>> createSnapshot(
             throw new IllegalStateException("Cannot create snapshot before the replica has been initialized");
         }
 
+        AbstractIterator<? extends RecordBatch> batchIterator = log.read(snapshotId.offset(), Isolation.COMMITTED).records.batchIterator();
+        if (batchIterator.hasNext() && batchIterator.peek().baseOffset() != snapshotId.offset()) {
+            logger.error("Cannot create snapshot at offset {} because it is not batch aligned", snapshotId.offset());
+            return Optional.empty();
+        }
```

Comment:
I know that I agreed to this solution in our offline discussion but since we are planning to change the UnifiedLog implementation to return the batch start offset in the LogOffsetMetadata, then maybe it is better to fix that PR (https://github.com/apache/kafka/pull/17528) first and merge it. After that we can fix this PR and Jira using the changes in the other PR.

Reply:
Agreed. Just checking the offset when it actually returns the batch start offset is more readable.